src/rgw/rgw_redis_scripts.lua

```diff
@@ -0,0 +1,231 @@
+    local name = keys[1]
+    local cookie = args[1]
+    local timeout = args[2]
+    local lock_status = assert_lock(keys, args)
+    if lock_status.map.errorCode == 0 then
+        redis.call('PEXPIRE', name, timeout)
+        return formatResponse(0, "", "")
+    elseif lock_status.map.errorCode == -lerrorCodes.ENOENT then
+        redis.call('SET', name, cookie, 'PX', timeout)
+        return formatResponse(0, "", "")
+    end
+    return lock_status
+end
+
+--- Release the lock on a resource.
+--- It deletes the key if the value matches the cookie.
+---@param keys table A single element list - lock name
+---@param args table A single-element list - cookie
+local function unlock(keys, args)
+    local name = keys[1]
+    local cookie = args[1]
+    local lock_status = assert_lock(keys, args)
+    if lock_status.map.errorCode == 0 then
+        redis.call('DEL', name)
+        return formatResponse(0, "", "")
+    end
+    return lock_status
+end
+
+--- Register the functions.
+redis.register_function('lock', lock)
+redis.register_function('unlock', unlock)
+redis.register_function('assert_lock', assert_lock)
+
+-----
+--- Queue functions
+-----
+
+-- Refer to 'Execution under low memory conditions'
+-- in https://redis.io/docs/latest/develop/interact/programmability/eval-intro/
+---
+
+-- Function to generate a string longer than n by concatenating the charset
+local function generateString(n)
+    local charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+    local result = {}
+    local charset_length = #charset
+
+    local repeats = math.ceil(n / charset_length) + 1
+    for i = 1, repeats do
+        table.insert(result, charset)
+    end
+    return table.concat(result)
+end
+
+--- Add an item to the reserve queue for n bytes
+--- @param keys table A single element list - queue name
+--- @param args table A single-element list - item size
+--- @return number 0 if the item is added to the queue, -lerrorCodes.ENOMEM if the queue is full
+local function reserve(keys, args)
+    local name = "reserve:" .. keys[1]
+    local item_size = tonumber(args[1])
+
+    local randomString = generateString(item_size)
+    --- generate a json of the format {"timestamp": <current time>, "data": <value>}
+    local time = redis.call("TIME")[0]
+    local value = '{"timestamp":' .. redis.call("TIME")[1] .. ',"data":"' .. randomString .. '"}'
+    if not redis.call('LPUSH', name, value) then
+        return formatResponse(-lerrorCodes.ENOMEM, "Not enough memory", "")
+    end
+    return formatResponse(0, "", "")
+end
+
+--- Remove an item from the reserve queue
+--- @param keys table A single element list - queue name
+--- @return number 0 if the item is removed from the queue, -lerrorCodes.ENOENT if the queue is empty
+local function unreserve(keys)
+    local name = "reserve:" .. keys[1]
+
+    local value = redis.call('RPOP', name)
+    if not value then
+        return formatResponse(-lerrorCodes.ENOENT, "Queue is empty", "")
```

Comment:
`ENOENT` indicate that the queue is deleted.
if there are no reservations in the queue, we can just assume `unreserve()` was successfull

Reply:
I think having logs for attempts to either commit or abort (which are the operations that call unreserve) a missing reservation would be good. 

What this would mean is that the operation took more time than the cleanup duration, is it good to have visibility for?

Or do you still think having it silent, i.e. reporting success for a missing reserve queue is fine?

Reply:
you are right, we should reply with an error.
however, we should not use `ENOENT` as this is an indications that the entire queue is deleted (not just missing a missing reservation),
the implications in our code for a deleted queue are different (we stop processign the queue).
in missing reservations we just log a warning

Reply:
you are right, we should reply with an error.
however, we should not use `ENOENT` as this is an indications that the entire queue is deleted (not just missing a missing reservation),
the implications in our code for a deleted queue are different (we stop processign the queue).
in missing reservations we just log a warning